src/axom/sina/include/Datum.hpp

```diff
@@ -0,0 +1,198 @@
+#ifndef SINA_DATUM_HPP
+#define SINA_DATUM_HPP
+
+/// @file
+
+#include <string>
+#include <type_traits>
+
+#include "axom/sina/include/ConduitUtil.hpp"
+#include "conduit.hpp"
+
+namespace axom
+{
+namespace sina
+{
+
+/**
+ * Represents whether a Datum is a String, Scalar (double), array of Strings,
+ * or array of Scalars.
+ */
+enum class ValueType {
+    String,
+    Scalar,
+    StringArray,
+    ScalarArray
+};
+
+/**
+ * A Datum tracks the value and (optionally) tags and/or units of a
+ * value associated with a Record, e.g. a scalar, a piece of metadata,
+ * or an input parameter. In the Sina schema, a Datum always
+ * belongs to a Record or one of Record's inheriting types.
+ *
+ * Every Datum must have a value; units and tags are optional.
+ *
+ * The value of a Datum may be a string, a double, an array of strings,
+ * or an array of doubles.
+ *
+ * \code
+ * axom::sina::Datum myDatum{12.34};
+ * std::string value = "foobar";
+ * axom::sina::Datum myOtherDatum{value};
+ * std::vector<double> scalars = {1, 2, 20.0};
+ * axom::sina::Datum myArrayDatum{scalars};
+ *
+ * //prints 1, corresponding to Scalar
+ * std::cout << static_cast<std::underlying_type<sina::ValueType>::type>(myDatum.getType()) << std::endl;
+ *
+ * //prints 0, corresponding to String
+ * std::cout << static_cast<std::underlying_type<sina::ValueType>::type>(myOtherDatum.getType()) << std::endl;
+ *
+ * //prints 3, corresponding to ScalarArray
+ * std::cout << static_cast<std::underlying_type<sina::ValueType>::type>(myArrayDatum.getType()) << std::endl;
+ *
+ * myRecord->add(myDatum);
+ * myOtherDatum.setUnits("km/s");
+ * myRecord->add(myOtherDatum);
+ * std::vector<std:string> tags = {"input", "core"};
+ * myArrayDatum.setTags(tags);
+ * myRecord->add(myArrayDatum);
+ * \endcode
+ */
+class Datum {
+public:
+    /**
+     * Construct a new Datum.
+     *
+     * @param value the string value of the datum
+     */
+    Datum(std::string value);
```

Comment:
"const std::string &" instead of std::string. This file should use references more for arguments so objects do not need to be copied to be passed as arguments.

Reply:
Currently this file is not pass by reference since we're concerned about things going out of scope during code cleanup. From @HaluskaR: "Since we never claim ownership of anything we're passed, we're at the mercy of codes not throwing things out...and they do like to do that. It makes us space inefficient but safe." This is probably worth discussing more during the meeting on Monday.

Reply:
Still better to pass by reference and then make your own copy. 

As it is now, there might be multiple copies.

Reply:
Same for the std::vectors as well.